### DISCUSSION TITLE
The State of Path Mapping

### DISCUSSION THREAD

**Discussion by fmeum** [bazelbuild/bazel#22658] [Show and tell]
## What is path mapping?

With path mapping enabled, Bazel automatically rewrites paths in action command lines with the aim of making them more likely to be disk or remote cache hits.
Specifically, configuration prefixes such as `bazel-out/darwin-amd64-fastbuild/bin` are replaced with a fixed string such as `bazel-out/cfg/bin`, so that the result of an action that doesn't depend on all aspects of the configuration encoded in the path (e.g. the OS, architecture and build mode) can be shared between different target platforms and configurations.

Path mapping by itself **won't** result in cache hits between different execution platforms because toolchains typically differ between platforms (in content, not just paths). One way to address this is to bundle toolchains for multiple execution platforms, see ["Universal toolchains"](https://docs.google.com/document/d/1uMPj2s0TlHSIKSngqOkWJoeqOtKzaxQLtBrRfYif3Lo/edit#heading=h.1nk1jojkq83g). Another is to use [cache key scrubbing](https://bazel.build/reference/command-line-reference#flag--experimental_remote_scrubbing_config).

For more information on path mapping, watch [the "Towards faster cross-platform builds" talk](https://www.youtube.com/watch?v=Et1rjb7ixUU).

## Using path mapping

Bazel has experimental support for path mapping for both native and Starlark rules as detailed below.

### General requirements

* At least Bazel 7.2.0rc3 or 8.0.0-pre.20240523.3 are recommended.
* Set `--experimental_output_paths=strip`.
* Use a disk or remote cache (or remote execution).
* Actions must use a compatible strategy:
  * `remote`
  * `dynamic`
  * `sandboxed`
  * `multiplex-worker` with `--experimental_worker_multiplex_sandboxing`
  * `worker` with `--worker_sandboxing` and `--noworker_multiplex`

## Java

Java compilation actions as well as header compilation actions support path mapping.

### Requirements

Since the default Java toolchain doesn't enable support for multiplex sandboxing yet, non-remote execution with path mapping requires using a custom Java toolchain:

```starlark
load("@bazel_tools//tools/jdk:default_java_toolchain.bzl", "default_java_toolchain")
default_java_toolchain(
    name = "java_toolchain",
    source_version = "17",
    target_version = "17",
    javac_supports_worker_multiplex_sandboxing = True,
)
```

rules_java 7.5.0 and higher have support for multiplex sandboxing.

## C/C++

C++ compilation and archiving actions have experimental support for path mapping.

### Requirements

Bazel 7.4.0 or higher is required and path mapping needs to be explicitly enabled for C/C++ compilation actions via `--modify_execution_info=CppCompile=+supports-path-mapping,CppModuleMap=+supports-path-mapping`. Support for `CppArchive` requires Bazel 8.1.0 or higher and requires opt-in in the same way.

## Go

rules_go optionally uses path mapping for its non-cgo compilation actions, including the stdlib.

### Requirements

rules_go from HEAD including [this commit](https://github.com/bazelbuild/rules_go/commit/6f206adcc9c2527a411a002669db1b7efafbeae8).

## Custom Starlark rules

Custom Starlark rules can opt into path mapping by adding an entry with the key `supports-path-mapping` to the `execution_requirements` dict passed to `ctx.actions.run` or `ctx.actions.run_shell`.
They can also be opted in from the command line via `--modify_execution_info`.

Additionally, command lines need to be constructed using `ctx.actions.args` and avoiding any of the methods on `File` that return a path string. Assuming that `src` is a `File` representing a file and `dir` is a `File` representing a directory ("tree artifact"):

1. instead of `args.add(src.path)`, use `args.add(src)`;
2. instead of `args.add(dir.path)`, use `args.add_all([dir], expand_directories = False)`;
3. instead of `args.add(src.dirname)`, use:
```starlark
def _dirname(file):
    return file.dirname

def _my_rule_impl(ctx):
    ...
    args.add_all([src], map_each = _dirname)
```

Generally speaking, all functions that return a path string that may contain a configuration prefix such as `bazel-out/darwin-amd64-fastbuild/bin` must only be called in `map_each` callbacks, where they are automatically path mapped by Bazel.

**Comment by illicitonion** [comment]
FWIW [rules_pkl](https://github.com/apple/rules_pkl) is a starlark rule which supports path mapping, see https://github.com/apple/rules_pkl/blob/bf3de08ea8c7f05329a2d59d9790ec8653d2a28c/pkl/private/pkl.bzl#L232 for an example.

**Reply by illicitonion** [reply]
When implementing support, we ran into https://github.com/bazelbuild/bazel/pull/16430#issuecomment-1833743749 - @fmeum I think the discussion there is currently waiting on a reply from you?

**Reply by fmeum** [reply]
I'll try to get back to it soon, but the main part of the discussion (what's the API going to look like, is `Args` supported, ...) is still very much open.

**Comment by lior10r** [comment]
Do you think that c++ support would be available at the next version (7.3.0)?

**Reply by fmeum** [reply]
I can't promise anything, but I would say that's very likely. The [main PR](https://github.com/bazelbuild/bazel/pull/22445) has already been approved and the follow-ups are pretty small.

**Reply by fmeum** [reply]
Basic support landed in 7.3.0, please see the updated main post

**Reply by lior10r** [reply]
Thanks! We started to try and working with it, at the moment our remote cache doesn't work. When we would manage to get I working I would let you know how it affected our build time.

**Reply by lior10r** [reply]
Hi, We started to work on integrating path mapping with our project and I have a question.
I saw that only the `CppCompile` supports path mapping and other cpp actions like `CppArchive` doesn't, is this true or I just didn't succeed to enable path mapping to this action? And if it really doesn't support path mapping is it planned to be supported in the future?

**Reply by fmeum** [reply]
This is correct, but adding support should be pretty simple. I will look into it for 8.1.0.

If this is a bottleneck for you, you could try switching to a linker with support for `--start-lib/--end-lib` to avoid building static libraries in the first place.

**Reply by lior10r** [reply]
Hi, I wanted to update that after using path mapping in our project we went from ~38m build time to ~6m!
We still have a lot of CppArchive duplicates which we think could make our build even faster. Do you think you would be able to get it into bazel 8.1.0?

**Reply by fmeum** [reply]
Thanks for the ping, I sent https://github.com/bazelbuild/bazel/pull/25081.

**Reply by gregestren** [reply]
@lior10r what kind of project do you have? That's great to hear such an improvement. How do you think path mapping produced that result for your project?

**Reply by lior10r** [reply]
We have a very big cpp project (~40000 targets), which uses a lot of transitions.
We compile multiple binaries, which use a lot of targets that are very similar, so the path mapping really helps.

**Reply by fmeum** [reply]
This will land in 8.1.0, I have updated the description.

**Comment by ThomasCJY** [comment]
Is there any plan to support android?

**Reply by fmeum** [reply]
Do you have a link to an example project I could try this on? It is being used for Android internally at Google, but I don't know the state of internal vs. external Android rules well enough to assess whether this translated into Bazel support.

**Reply by gregestren** [reply]
Hi @ThomasCJY. Our original focus with path mapping was equally on Java / Android. We experimented with supporting Android actions but would likely need to update them to conform to the current path mapping API. Happy to look into this, and especially run over any examples if you have any.

**Reply by ThomasCJY** [reply]
I created a simple test app which failed to build with path mapping enabled. 

Code: https://github.com/ThomasCJY/testApp/tree/jchen-testPathMapping

You can repro the issue by running `bazelisk clean; bazelisk build //app/src:my_application`

The error log shows:
```
ERROR: /private/var/tmp/_bazel_jchen/d841f813fb03211ada45ecca382b434a/external/bazel_tools/tools/android/BUILD:43:12: output 'external/bazel_tools/tools/android/d8_dexbuilder.jar' was not created
ERROR: /private/var/tmp/_bazel_jchen/d841f813fb03211ada45ecca382b434a/external/bazel_tools/tools/android/BUILD:43:12: output 'external/bazel_tools/tools/android/d8_dexbuilder-native-header.jar' was not created
ERROR: /private/var/tmp/_bazel_jchen/d841f813fb03211ada45ecca382b434a/external/bazel_tools/tools/android/BUILD:43:12: output 'external/bazel_tools/tools/android/d8_dexbuilder.jar_manifest_proto' was not created
ERROR: /private/var/tmp/_bazel_jchen/d841f813fb03211ada45ecca382b434a/external/bazel_tools/tools/android/BUILD:43:12: Building external/bazel_tools/tools/android/d8_dexbuilder.jar () [for tool] failed: not all outputs were created or valid
```
which indicates some d8 related issue

**Reply by fmeum** [reply]
Did you make the changes required for path mapping for Java? Those would be a requirement for this to work and it looks like the failure is caused by a Java compilation action not being path mapped properly.

**Reply by ThomasCJY** [reply]
Yes, I pushed the updates on my test repo https://github.com/ThomasCJY/testApp/tree/jchen-testPathMapping which registered the custom java toolchain and it still failed with the same error

**Reply by ThomasCJY** [reply]
@fmeum did you get any chance to test it out?

**Reply by fmeum** [reply]
I tried to get it to work, but I didn't manage to get to this point: Android Studio fails to install build-tools 30.0.2 for me, but when I set `build_tools_version = "31.0.0"`, the Bazel build still fails with references to `@android_sdk//build-tools/30.0.2:aapt2` missing, even after `bazel clean --expunge`. Do you know how I can get them updated?

That said, I think that you are missing a flag (the description of this discussion didn't make it clear that you need it): try adding `--noworker_multiplex`.

**Reply by ThomasCJY** [reply]
yeah you are right. Adding `--noworker_multiplex` fixed the build issue. Any idea why I am still seeing `darwin_arm64-xxx` directories under `bazel-out` while no `bazel-out/cfg/` is generated?
```
ls bazel-out/
_tmp
darwin_arm64-opt-exec-ST-d57f47055a04
darwin_arm64-fastbuild
darwin_arm64-fastbuild-android-ST-5b74a929aefd
```

**Reply by fmeum** [reply]
This is expected and actually a feature: path mapping only applies at execution time to ensure that downstream actions see all the files in the expected locations, regardless of whether they use path mapping themselves. Try running with `-s` to see the mapped paths.

**Comment by fmeum** [comment]
Edit: Path mapping in 7.3.0 is currently broken due to race conditions in the new action deduplication functionality. See https://github.com/bazelbuild/bazel/issues/23288.

**Reply by fmeum** [reply]
This has been hotfixed in 7.3.1 by disabling local execution deduplication. I have a full fix out in https://github.com/bazelbuild/bazel/pull/23382. 

Performance data obtained with the patch on Bazel itself with a suitable Java toolchain:
```
# Baseline
INFO: Elapsed time: 102.879s, Critical Path: 24.86s
INFO: 5128 processes: 145 internal, 3076 darwin-sandbox, 1907 worker.

# --experimental_worker_multiplex_sandboxing
INFO: Elapsed time: 115.551s, Critical Path: 29.83s
INFO: 5128 processes: 145 internal, 3076 darwin-sandbox, 1907 worker.

# --experimental_worker_multiplex_sandboxing --experimental_output_paths=strip, without deduplication
INFO: Elapsed time: 95.024s, Critical Path: 27.01s
INFO: 5128 processes: 1438 disk cache hit, 145 internal, 2375 darwin-sandbox, 1170 worker.

# --experimental_worker_multiplex_sandboxing --experimental_output_paths=strip, with deduplication
INFO: Elapsed time: 80.242s, Critical Path: 22.04s
INFO: 5128 processes: 1367 disk cache hit, 145 internal, 2171 darwin-sandbox, 459 deduplicated, 986 worker.
```

**Comment by sputt** [comment]
Just started prototyping this and the results so far are amazing.  Subtle configuration differences no longer cause duplication of super expensive actions.  For me it is generated C files and architectureless data files.  Thanks for doing this!  I feel like we're just beginning to recognize the impact and will find much more savings.

Deduplicating executions is important for us since local executions of these duplicates often get scheduled around the same time.  Another interesting note is the ability to cap disk cache size might be required before we roll this out to workstations.

**Reply by fmeum** [reply]
Action deduplication will land in 7.4.0 (https://github.com/bazelbuild/bazel/pull/23674).

**Comment by nikhilkalige** [comment]
@fmeum As I am trying to fix my rules to support path mapping, I wanted to see what rules could possibly be fixed to use this. I wanted to see all the cmd line args of a particular target, to help me figure out the rules that need fixing. I tried `aquery` and in that, I don't see the arguments post mapping, which kind of makes sense. So, do you know of a way to get the cmd line post mapping somehow?

**Reply by fmeum** [reply]
The only realistic way to see what Bazel actually executes is to run a build with `-s`. You could set up a `--disk_cache` and build after a `bazel clean`, which will still be very fast due to caching but show you all the command lines.

**Reply by nikhilkalige** [reply]
That's exactly what I have been doing.. Thank you..

**Comment by nikhilkalige** [comment]
We have a few rules that take directory as an argument. Most locations kind used this
```py
args.add(file.dirname)
# Updated now to
args.add_all([file], map_each = _dirname)
```
Above suffices in a lot of places, but becomes very quirky when you have complex rules. For example, in protobuf rules, there is a lot of directory references passed to `-I`. 

- Is it possible to have an api to construct a `path` from starlark?
- `File.dirname` return a `path` instead of `string`, would also be useful

**Reply by fmeum** [reply]
So far we haven't done this since path mapping *may* (but currently doesn't) map different files in the same directory to different output roots, so `file.dirname` returning a `path` still wouldn't have enough information to path map. When we get to the point that we decide that a static output root (`bazel-out/cfg/bin`) is sufficient, we could start adding better API for this use case.

@gregestren @aranguyen Are you still interested in exploring content-based path mapping schemes?

**Reply by gregestren** [reply]
> Are you still interested in exploring content-based path mapping schemes?

Was there a doc reviewing this or did we only discuss this in PR reviews / chats?

**Reply by fmeum** [reply]
I only know what's in https://docs.google.com/document/d/17snvmic26-QdGuwVw55Gl0oOufw9sCVuOAvHqGZJFr4/edit and code comments.

**Reply by gregestren** [reply]
Am I reading this thread correctly that exploring content-based approaches is an obstacle to this request, since it forks the directories of an action's inputs? 

Thanks for reminding me about  https://docs.google.com/document/d/17snvmic26-QdGuwVw55Gl0oOufw9sCVuOAvHqGZJFr4/edit. :) A few interesting ideas we could further explore:

- Adding smarter `cfg` distinguishers to handle multi-configured actions. That's still unsupported, right?
- Applying the mapping to Bazel's output tree instead of just the executor (which the above doc describes). That can better guarantee all consumers of action outputs (IDEs, post-processing shell scripts, ...) have the same view as the executor that ran the action 
- If we went all-in making the local output tree a CAS with some mapping API to realize "important" outputs, we could avoid action conflicts by design. In other words, break the pattern that Bazel needs distinct artifact paths over the scope of the entire output tree. Instead paths only need to be unique over the scope of what outputs the build materializes. I'm hand-waving here because ideally this also applies inside the Bazel process itself, but that's a whole new level of complication.

I don't know - what are your current thoughts? I remember you were interested in exploring more sophisticated mapping algorithms. Anything particular still on your mind? As much as I think the above ideas are fascinating, at some point they become major engineering work and I wonder if the current simple model is good enough for most practical use cases.

**Reply by fmeum** [reply]
Sorry, I completely missed your reply. 

I currently don't see a use case in which more sophisticated mapping algorithms would be worth the added complexity. Even for actions that embed output paths (e.g. C++ debug symbols) you typically wouldn't mix different configurations, so materializing `bazel-out/cfg` as symlinks would still work. I would rather try that first then go with a more complex scheme.

**Comment by keith** [comment]
I was testing this out with https://github.com/bazelbuild/bazel/pull/23630 applied, I got pretty far but I hit this issue:

```
ERROR: AsyncRT/BUILD.bazel:138:23: Compiling AsyncRT/lib/Runtime/Globals/Globals.cpp failed: undeclared inclusion(s) in rule '//AsyncRT:RuntimeGlobals.static':
this rule is missing dependency declarations for the following files included by 'AsyncRT/lib/Runtime/Globals/Globals.cpp':
  'bazel-out/cfg/bin/external/gperftools/tcmalloc/include/gperftools/tcmalloc.h'
```

Where the header here is from a rules_foreign_cc target:

```
configure_make(
    name = "tcmalloc",
    autogen = True,
    configure_in_place = True,
    configure_options = [
        "--enable-shared=no",
    ],
    copts = [
        "-fPIC",
        "-w",
        "-fexceptions",  # Needed for tests, which ideally we wouldn't build
        "-Wno-register",
    ],
    lib_source = ":all_srcs",
    out_static_libs = ["libtcmalloc_minimal.a"],
    target_compatible_with = select({
        "@platforms//os:linux": [],
        "//conditions:default": ["@platforms//:incompatible"],
    }),
    visibility = ["//visibility:public"],
)
```

It might be relevant that I access it through an alias (although removing that doesn't help):

```
alias(
    name = "tcmalloc",
    actual = select({
        "@platforms//os:macos": "@gperftools-macos//:tcmalloc",
        "@platforms//os:linux": "@gperftools//:tcmalloc",
    }),
)
```

Are you tracking this one?

**Reply by fmeum** [reply]
I am now: https://github.com/bazelbuild/bazel/issues/23681
I have an idea what could be wrong, will try to reproduce it in an integration test.

**Comment by jjudd** [comment]
I've been implementing path mapping in few rule sets and I think I'm either implementing something incorrectly or I've found an issue with path mapping + multiplex worker sandboxing. I'm seeing errors where the worker doesn't get run because the executable isn't present in the sandbox. `IOException while preparing the execution environment of a worker`

Both path mapping and multiplex sandboxing work on their own:

- Disabling multiplex workers with `--noworker_multiplex` gets path mapping working.
- Disabling path mapping with `--experimental_output_paths=off` gets multiplex sandboxing working.

Here's a minimal repro case as well as the full error (in the README): https://github.com/lucidsoftware/path-mapping-bug-repro You can repro with `bazel build example`. The worker doesn't do anything, but it doesn't need to in order to repro the bug as we're not getting to the point where the worker actually runs.

**Reply by fmeum** [reply]
I think that this is happening because your worker binary is a generated file. That's not a case I have tested so far since the Java worker is a precompiled deploy jar. I will look into this.

**Reply by jjudd** [reply]
Thank you very much. Greatly appreciated.

**Reply by fmeum** [reply]
I have a fix: https://github.com/bazelbuild/bazel/pull/23991

**Comment by jadenPete** [comment]
I'm attempting to implement path mapping in a ruleset, and I'm not sure if I'm misunderstanding how path mapping works, but I'm not able to get cache hits between what should be identical actions. I think I may have discovered a bug.

Here's a repository that minimally reproduces the issue:
https://github.com/lucidsoftware/path-mapping-action-key-bug-repro

The `README.md` explains the issue and running `example/test` should demonstrate it.

Do you know why this is happening and how I can prevent `//example:file` from being built twice?

**Reply by fmeum** [reply]
You need to use a disk or remote cache to see any benefits from path mapping. This is because as you discovered, the *actions* defined by the two rules are still distinct, but the *command invocations* they result in (which Bazel internally refers to as "spawns") are identical. 

You can confirm this by running with `-s` and observing that the executed command lines are the same, which means that either will be a (disk or remote) cache hit for the other. Since these spawns are very likely to be run concurrently, the deduplication feature added in 7.4.0 will kick in and give you a hit even before any of the spawns have uploaded their result to the cache. On your example, with `--disk_cache=some/path`, this looks as follows:

```
Starting local Bazel server and connecting to it...
INFO: Invocation ID: 9785dcd9-9f2f-43c9-b96b-f3e6ddae5851
INFO: Analyzed 4 targets (6 packages loaded, 14 targets configured).
INFO: From StaticFile example/file/output.txt:
Generating static file...
INFO: From StaticFile example/file/output.txt:
Generating static file...
INFO: Found 4 targets...
INFO: Elapsed time: 4.337s, Critical Path: 0.09s
INFO: 5 processes: 1 internal, 3 darwin-sandbox, 1 deduplicated.
INFO: Build completed successfully, 5 total actions
```

**Reply by jadenPete** [reply]
I can confirm that makes it work. I think this was a misunderstanding of how Bazel does caching on my part.
Thank you for your help!

**Comment by keith** [comment]
Filed some tracking issues for some know issues:

https://github.com/bazelbuild/bazel/issues/25359
https://github.com/bazelbuild/bazel/issues/25360

**Comment by jjudd** [comment]
I think I found an issue with path mapping. Either that or I'm doing something wrong. I have a minimal repro case here: https://github.com/lucidsoftware/path-mapping-bug-2-repro

Enabling pathmapping for the workspace and disabling sandboxing on a target via the `"no-sandbox"` tag causes the action to fail in unintuitive ways for some rules.

*Problem*
The repo has a simple rule implemented in `example/example_rule.bzl` that wraps singlejar. That action supports pathmapping and sets the `"supports-path-mapping": "1"` execution requirement.

The repo sets up a custom Java toolchain, so path mapping will work with the Java rules.

When pathmapping + sandboxing are enabled, things work.

When pathmapping is enabled and the `no-sandbox` tag is set on `//example:example`, things break as follows:
`singlejar_local: src/tools/singlejar/output_jar.cc:322: bazel-out/cfg/bin/example/foo.jar: No such file or directory`

It seems that all paths passed to the singlejar command are not properly mapped. This is the command line that results in the error above. Note the `bazel-out/cfg/bin` in the paths.
```external/rules_java++toolchains+remote_java_tools_linux/java_tools/src/tools/singlejar/singlejar_local --output bazel-out/cfg/bin/example/foo.jar --sources bazel-out/cfg/bin/example/libexample_foo.jar```

*Oddity*
The `no-sandbox` tag works fine on the `//example:example_java_library` depended upon by `//example:example`, so this doesn't seem to be a problem with `java_library`

**Reply by fmeum** [reply]
This can be explained, but I agree it's not a great state:
* Path mapping requires some kind of sandboxing, be it remote, sandboxed worker or regular sandboxed execution. This is pretty much impossible to change, we need some level of indirection to enable the path rewriting.
* `no-sandbox` only disables sandboxed execution, *not* sandboxed worker execution. This is why `no-sandbox` on your example rule has it fall back to unsandboxed local execution, with the failure mode you observed, while it still allows `java_library` to use the sandboxed multiplex worker execution strategy, which does work with path mapping. This doesn't make much sense, but changing it could have backwards compatibility implications. FYI @bigelephant29, is this something you think we could change?
* Strategy selection currently doesn't take path mapping into account and may attempt to run a path mapped spawn in an unsandboxed environment, where it fails in confusing ways. I have https://github.com/bazelbuild/bazel/pull/25430 out to fail strategy selection in that case instead.

**Reply by jjudd** [reply]
Sorry for the late response. Thanks for your response, much appreciated.
* Totally understand that a sandbox is required for path mapping.
* My biggest concern is that it was an obscure execution error rather than something more obvious. Your PR to skip strategies incompatible with path mapping would likely have been far easier to understand.
* I did not know that `no-sandbox` only disabled sandboxed execution and not sandboxed worker execution. TIL. That seems like a footgun, but I also get that changing it may break some folks who are depending on that behavior.

**Reply by bigelephant29** [reply]
I think this is WAI and we'd like to keep it as is. If you'd like to disable sandbox worker execution, the only way to do it now is to disable `experimental_worker_multiplex_sandboxing`, which is enabled in your blazerc. cc @zhengwei143
